types/eslint/index.d.ts

```diff
@@ -431,85 +433,107 @@ export namespace SourceCode {
-        MemberExpression?: ((node: ESTree.MemberExpression & NodeParentExtension) => void) | undefined;
-        MetaProperty?: ((node: ESTree.MetaProperty & NodeParentExtension) => void) | undefined;
-        MethodDefinition?: ((node: ESTree.MethodDefinition & NodeParentExtension) => void) | undefined;
-        NewExpression?: ((node: ESTree.NewExpression & NodeParentExtension) => void) | undefined;
-        ObjectExpression?: ((node: ESTree.ObjectExpression & NodeParentExtension) => void) | undefined;
-        ObjectPattern?: ((node: ESTree.ObjectPattern & NodeParentExtension) => void) | undefined;
-        Program?: ((node: ESTree.Program) => void) | undefined;
-        Property?: ((node: ESTree.Property & NodeParentExtension) => void) | undefined;
-        RestElement?: ((node: ESTree.RestElement & NodeParentExtension) => void) | undefined;
-        ReturnStatement?: ((node: ESTree.ReturnStatement & NodeParentExtension) => void) | undefined;
-        SequenceExpression?: ((node: ESTree.SequenceExpression & NodeParentExtension) => void) | undefined;
-        SpreadElement?: ((node: ESTree.SpreadElement & NodeParentExtension) => void) | undefined;
-        Super?: ((node: ESTree.Super & NodeParentExtension) => void) | undefined;
-        SwitchCase?: ((node: ESTree.SwitchCase & NodeParentExtension) => void) | undefined;
-        SwitchStatement?: ((node: ESTree.SwitchStatement & NodeParentExtension) => void) | undefined;
-        TaggedTemplateExpression?: ((node: ESTree.TaggedTemplateExpression & NodeParentExtension) => void) | undefined;
-        TemplateElement?: ((node: ESTree.TemplateElement & NodeParentExtension) => void) | undefined;
-        TemplateLiteral?: ((node: ESTree.TemplateLiteral & NodeParentExtension) => void) | undefined;
-        ThisExpression?: ((node: ESTree.ThisExpression & NodeParentExtension) => void) | undefined;
-        ThrowStatement?: ((node: ESTree.ThrowStatement & NodeParentExtension) => void) | undefined;
-        TryStatement?: ((node: ESTree.TryStatement & NodeParentExtension) => void) | undefined;
-        UnaryExpression?: ((node: ESTree.UnaryExpression & NodeParentExtension) => void) | undefined;
-        UpdateExpression?: ((node: ESTree.UpdateExpression & NodeParentExtension) => void) | undefined;
-        VariableDeclaration?: ((node: ESTree.VariableDeclaration & NodeParentExtension) => void) | undefined;
-        VariableDeclarator?: ((node: ESTree.VariableDeclarator & NodeParentExtension) => void) | undefined;
-        WhileStatement?: ((node: ESTree.WhileStatement & NodeParentExtension) => void) | undefined;
-        WithStatement?: ((node: ESTree.WithStatement & NodeParentExtension) => void) | undefined;
-        YieldExpression?: ((node: ESTree.YieldExpression & NodeParentExtension) => void) | undefined;
-    }
+        meta?: RuleMetaData;
+    }
+
+    type MixedNode = ESTree.Program | NodeTypeMap[keyof NodeTypeMap];
+
+    type NodeTypes = MixedNode["type"];
+
+    interface NodeTypeMap {
```

Comment:
allow user declaration merging add more ast node, e.g: typescript-eslint

Reply:
The problem with this is that it requires the "extension" to redeclare every single node in the AST that uses specific unions.

For any non-trivial extension this means a complete redefinition of the AST.